Fix codegen of consteval functions returning an empty class, and related issues

efriedma-quic opened the pull request:
Fix codegen of consteval functions returning an empty class, and related issues

If a class is empty, don't store it to memory: the store might overwrite useful data.  Similarly, if a class has tail padding that might overlap other fields, don't store the tail padding to memory.

The problem here turned out a bit more general than I initially thought: basically all uses of EmitAggregateStore were broken. Call lowering had a method that did mostly the right thing, though: CreateCoercedStore. Adapt CreateCoercedStore so it always does the conservatively right thing, and use it for both calls and ConstantExpr.

Also, along the way, fix the "overlap" bit in AggValueSlot: the bit was set incorrectly for empty classes in some cases.

Fixes #93040.

zygoloid reviewed (approved):
Looks good. Do we also need to worry about overwriting tail padding here?

efriedma-quic commented:
I didn't think so at first glance... but yes, we do, in certain obscure cases:

```
#include <new>
struct A { char c; A(); };
struct __attribute((packed)) S  { char a; int x; __attribute((aligned(2))) char y; consteval S() : x(1), a(3), y(2) {} };
struct S2 { [[no_unique_address]] S s; [[no_unique_address]] A a; };
static_assert(sizeof(S)==8 && sizeof(S2)==8);
void f2(S2 *s) { new (&s->s) S; }
```

I'll look into reworking this.

serge-sans-paille commented on clang/lib/CodeGen/CGExprAgg.cpp:
```diff
@@ -135,6 +135,17 @@ class AggExprEmitter : public StmtVisitor<AggExprEmitter> {
 
     if (llvm::Value *Result = ConstantEmitter(CGF).tryEmitConstantExpr(E)) {
+      // An empty record can overlap other data (if declared with
+      // no_unique_address); omit the store for such types - as there is no
```
Candide question: Empty record still need one byte when their address is taken (thus this comment about `no_unique_address` I guess), why don't we see that in the diff?

efriedma-quic replied:
See what, exactly?  Given the derived class, computing the address of the base class doesn't take any instructions, because it's the same address.

efriedma-quic committed:
Fix codegen of consteval functions returning an empty class, and related issues

If a class is empty, don't store it to memory: the store might overwrite
useful data.  Similarly, if a class has tail padding that might overlap
other fields, don't store the tail padding to memory.

The problem here turned out a bit more general than I initially thought:
basically all uses of EmitAggregateStore were broken. Call lowering had
a method that did mostly the right thing, though: CreateCoercedStore.
Adapt CreateCoercedStore so it always does the conservatively right
thing, and use it for both calls and ConstantExpr.

Also, along the way, fix the "overlap" bit in AggValueSlot: the bit was
set incorrectly for empty classes in some cases.

Fixes #93040.

efriedma-quic commented on clang/test/CodeGenOpenCL/addr-space-struct-arg.cl:
```diff
@@ -177,7 +179,12 @@ kernel void KernelTwoMember(struct StructTwoMember u) {
 // AMDGCN-SAME:  (%struct.LargeStructTwoMember %[[u_coerce:.*]])
 // AMDGCN:  %[[u:.*]] = alloca %struct.LargeStructTwoMember, align 8, addrspace(5)
-// AMDGCN:  store %struct.LargeStructTwoMember %[[u_coerce]], ptr addrspace(5) %[[u]]
+// AMDGCN:  %[[U_PTR0:.*]] = getelementptr inbounds %struct.LargeStructTwoMember, ptr addrspace(5) %[[u]], i32 0, i32 0
```
Unifying the codepaths makes FCA promotion happen more often.

efriedma-quic commented on clang/test/CodeGenCXX/address-space-cast-coerce.cpp:
```diff
@@ -46,9 +46,9 @@ int mane() {
     char1 f1{1};
     char1 f2{1};
 
-// CHECK: [[TMP:%.+]] = alloca i16
```
The revised version of casting integers is a bit more aggressive; it's hard to make it precisely match the old code while still preserving the correct semantics.

efriedma-quic commented on clang/test/CodeGen/arm-mve-intrinsics/vld24.c:
```diff
@@ -48,10 +48,13 @@ uint8x16x4_t test_vld4q_u8(const uint8_t *addr)
-// CHECK-NEXT:    [[VALUE_COERCE_FCA_0_1_EXTRACT:%.*]] = extractvalue [[STRUCT_UINT32X4X2_T]] [[VALUE_COERCE]], 0, 1
-// CHECK-NEXT:    call void @llvm.arm.mve.vst2q.p0.v4i32(ptr [[ADDR:%.*]], <4 x i32> [[VALUE_COERCE_FCA_0_0_EXTRACT]], <4 x i32> [[VALUE_COERCE_FCA_0_1_EXTRACT]], i32 0)
-// CHECK-NEXT:    call void @llvm.arm.mve.vst2q.p0.v4i32(ptr [[ADDR]], <4 x i32> [[VALUE_COERCE_FCA_0_0_EXTRACT]], <4 x i32> [[VALUE_COERCE_FCA_0_1_EXTRACT]], i32 1)
+// CHECK-NEXT:    [[TMP0:%.*]] = extractvalue [[STRUCT_UINT32X4X2_T:%.*]] [[VALUE_COERCE:%.*]], 0
```
Apparently I've stumbled over some limitation of instcombine.

efriedma-quic commented on clang/test/CodeGen/arm-vfp16-arguments2.cpp:
```diff
@@ -44,20 +44,20 @@ struct S1 f1(struct S1 s1) { return s1; }
 // CHECK-SOFT: define{{.*}} void @_Z2f22S2(ptr dead_on_unwind noalias nocapture writable writeonly sret(%struct.S2) align 8 %agg.result, [4 x i32] %s2.coerce)
 // CHECK-HARD: define{{.*}} arm_aapcs_vfpcc [2 x <2 x i32>] @_Z2f22S2([2 x <2 x i32>] returned %s2.coerce)
-// CHECK-FULL: define{{.*}} arm_aapcs_vfpcc %struct.S2 @_Z2f22S2(%struct.S2 returned %s2.coerce)
+// CHECK-FULL: define{{.*}} arm_aapcs_vfpcc %struct.S2 @_Z2f22S2(%struct.S2 %s2.coerce)
```
This is also the instcombine issue.

efriedma-quic commented:
(I'd like a re-review of the latest version: I made significant revisions to address the tail-padding issues.)

zygoloid commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -1336,75 +1336,56 @@ static llvm::Value *CreateCoercedLoad(Address Src, llvm::Type *Ty,
-    Dst = Dst.withElementType(SrcTy);
-    CGF.EmitAggregateStore(Src, Dst, DstIsVolatile);
+  if (SrcSize.isScalable() || SrcSize <= DstSize) {
+    if (SrcTy->isIntegerTy() && Dst.getElementType()->isPointerTy() &&
+        SrcSize == CGM.getDataLayout().getTypeAllocSize(Dst.getElementType())) {
+      // If the value is supposed to be a pointer, convert it before storing it.
+      Src = CoerceIntOrPtrToIntOrPtr(Src, Dst.getElementType(), *this);
```
It looks like we used to create either an `inttoptr` or `ptrtoint` here and now we only handle the `inttoptr` case. Is that OK?

efriedma-quic replied:
inttoptr comes from constructs like a struct containing a pointer, which the calling convention code sometimes chooses to just coerce to an integer instead of trying to compute a prettier type.  The other direction would be very unusual: it would mean we somehow coerced some non-pointer value to a pointer.

And this isn't really a correctness thing, anyway; type-punning in memory is basically the same as doing it in a register.  This just makes the intended types more clear for optimizations.

zygoloid replied:
Yeah, it definitely seems like it'd be a pretty big surprise to need this. OK.

zygoloid commented on clang/lib/CodeGen/CodeGenFunction.h:
```diff
@@ -4772,7 +4772,8 @@ class CodeGenFunction : public CodeGenTypeCache {
 
   /// Build all the stores needed to initialize an aggregate at Dest with the
   /// value Val.
```
This comment looks out of date.

RSilicon commented:
Is this worth back porting as it is a bugfix over code gen?

AaronBallman commented:
> Is this worth back porting as it is a bugfix over code gen?

IMO, it's worth considering, but if we want to go down this route, I think we need to do so relatively quickly -- we have about two weeks until rc3, and given the size of this change, I'm not certain we should try landing it any later than rc3 just due to risk.

efriedma-quic commented:
This is maybe slightly risky in terms of possible regressions, but it is a fix for a miscompile, and we're early enough in the release process that it's probably fine.

/cherry-pick 1762e01

efriedma-quic commented:
/cherry-pick 1762e01cca0186f1862db561cfd9019164b8c654
